### DISCUSSION TITLE
Add `onPrefetch` to `Link`

### DISCUSSION THREAD

**Discussion by ryanflorence** [remix-run/react-router#12375] [Proposals]
We can prefetch loader data because we simply use html `<link rel="prefetch" />` and then the browser manages the cache with HTTP. We get feature requests to prefetch client loaders too but then we'd need a cache to store them (and expire them) etc.

Instead, I think we can add a very simple event to links: `onPrefetch`.

Apps can simply do whatever they'd like there, probably call `prefetch` on whatever cache they're using (like React Query).

```tsx
// client loader uses fetchQuery already
export async function clientLoader({ params }) {
  return queryClient.fetchQuery({
    queryKey: ["notes", params.id],
    queryFn: getNotes,
  });
}

// links can prefetch anything they need to
<Link
  to="/notes"
  prefetch="intent"
  onPrefetch={() => {
    queryClient.prefetchQuery({
      queryKey: ["notes", "abc"],
      queryFn: getNotes,
    });
  }}
/>;
```

This way we can still prefetch the JS modules and they can prefetch everything else.

## Signature

I don't think it needs any arguments because they should have everything they need in scope already:

```tsx
notes.map(note => (
  <Link
    to={note.id}
    prefetch="intent"
    onPrefetch={(eventStub) => {
      if (whatever) eventStub.preventDefault()
      queryClient.prefetchQuery({
        // know the ID already, don't need params
        queryKey: ["notes", id],
        queryFn: getNotes,
      });
    }}
  />
));
```

**Comment by jgoz** [comment]
Within data routers (library mode), it would be super cool if the default prefetch behavior was to invoke the loader for the target href. `onPrefetch`, if defined, would take precedence.

This would encapsulate data loading particulars within the route and encourage the pattern of fetching/prefetching in the loader to prime the cache, closer to where the query results are used.

But this could totally be a userland thing too:

```ts
function useLinkPrefetcher({ to, prefetch }: Pick<LinkProps, 'to' | 'prefetch'>, key?: string) {
  const href = useHref(to);
  const { load } = useFetcher({ key: key ?? `preload/${to}` });

  return useMemo(() => {
    return prefetch ? () => load(href) : undefined;
  }, [href, load, prefetch]);
}
```

**Comment by dantrain** [comment]
Great that this is being proposed!

> they should have everything they need in scope already

I don't think this is necessarily a safe assumption. There could be things in scope in the route module available to be used in the client loader that are not in scope everywhere a `<Link>` is used.

It would be great to have the option to call the client loader if desired, for instance if `queryClient.prefetchQuery()` is sufficient for both navigation and prefetching cases.

**Comment by brophdawg11** [comment]
Related/Duplicate Proposal: https://github.com/remix-run/react-router/discussions/14187

**Comment by brophdawg11** [comment]
We are going to move forward with this.  Adding the `accepting prs` label if anyone from the community wants to take a stab at this, but otherwise one of us will try to pick this up before long.

**Reply by jacob-ebey** [reply]
👍

**Reply by markdalgleish** [reply]
👍

**Reply by jin-Pro** [reply]
@brophdawg11 

I was the one who reported issue #14187.

I want not only Link, but also Routes definitions to support 'relations' or 'prefetch', so that when a page is loaded, the related lazy components are loaded together.

I wanted to contribute this feature myself, but since my skills aren’t quite there yet, I’ll just keep an eye on it for now.

**Reply by brophdawg11** [reply]
You can already handle page component load behaviors with `route.lazy`?  

If we moved this to the route level, then there would need to be an additional required route matching step before we could call `onPrefetch` which wouldn't be necessary in all cases if you already know exactly what prefetch logic you wanted to run.  Putting it on the Link also allows it to be conditional more easily, and avoids having to pass parameters to a route.prefetch() handler that you need to then conditionally prefetch based on.

FWIW, I think this API will allow you to do what you're looking for in user land too:

```tsx
let reouter = creteBrowserRouter([{
  path: '/',
  loader() { ... },
  handle: {
    prefetch() { ... }
  }
}]);

// Then in your app
<Link 
  to="/path" 
  prefetch="intent" 
  onPrefetch={() => matchRoutes(router.routes)?.forEach(m => m.handle?.prefetch())}
>
```

**Comment by brophdawg11** [comment]
Related proposal closed in favor of this one https://github.com/remix-run/react-router/discussions/8012

**Comment by ariesclark** [comment]
## Suggestion: Route-level preloading.

Rather than placing preloading logic in the Link component, I believe this functionality should be exported from routes themselves. This ensures that every reference to a route automatically preloads, eliminating the need to wrap React Router's Link component in userland.

Here's an implementation using the https://reactrouter.com/start/framework/route-module#handle:

In your route file:
  ```tsx
  export const handle = {
    async intent() {
      // do anything.
    }
  };

  export default function Component() { /* ... */ }
  ```

  In your link component:
  ```tsx
  import { Link, matchRoutes } from "react-router";

  function IntentfulLink({ to, children }) {
    let intent = React.useCallback(async () => {
      // @ts-expect-error: React Router internal.
      const matches = matchRoutes(globalThis.__reactRouterDataRouter.routes, to) || [];

      await Promise.all(matches.map(async ({ route }) => {
        // @ts-expect-error: React Router internal.
        const handle = route.handle || await route.lazy?.handle();
        if (!handle || typeof handle !== "object" || !handle.intent) return null;

        return handle.intent();
      }));
    }, [to]);

    return (
      <Link to={to} onMouseOver={intent}>
        {children}
      </Link>
    );
  }
```

**Reply by brophdawg11** [reply]
We should not assume that users want automatic preloading.  As mentioned [above](https://github.com/remix-run/react-router/discussions/12375#discussioncomment-14180792), the more granular `Link` API we are proposing allows fine grained preloading as well as automatic preloading with a small user-land wrapper.

**Comment by akrynski-1** [comment]
Will there be an option to simply call the loader? I think many tanstack query users (including me) would appreciate not having to duplicate data loading code

**Reply by brophdawg11** [reply]
I think that would be a good idea, the trick is whether we need to let you know that it's a prefetch execution in case you need to behave differently - I don't know if the majority of cases will behave _identically_ in the loader when prefetching versus actual fetching so I don't think we want to make that assumption.

We could potentially design it similar to framework mode's `clientLoader` which receives a `serverLoader` function to make the request to the server loader:

```tsx
let router = createBrowserRouter([{
  path: '/',
  async loader() {
    let cachedData = cache.get(cacheKey);
    if (cachedData) { return cachedData; }

    let data = await getExpensiveData();
    cache.set(cacheKey, data);
    return data;
  },
  Component: Index,
}]);

// Then in your app
<Link to="/" prefetch="intent" onPrefetch={async ({ request, params, context, loader }) => { 
  // Call the loader verbatim
  await loader();
}}>
```

Maybe we could just let folks use `context` to indicate that it's a prefetch:

```tsx
let prefetchContext = createContext(false);

let router = createBrowserRouter([{
  path: '/',
  async loader({ context }) {
    if (context.get(prefetchContext)) {
      // Do something differently when it's a prefetch invocation
    } else {
      return await getData();
    }
  },
  Component: Index,
}]);

// Then in your app
<Link to="/" prefetch="intent" onPrefetch={async ({ context, loader }) => { 
  context.set(prefetchContext, true);
  await loader();
}}>
```

**Reply by jgoz** [reply]
Being able to call the loader would be amazing. I have a similar use case where the tanstack query cache is primed via prefetches in the loader and have it wired up to a custom prefetch=intent implementation.

The hardest and most brittle part was matching a target href to a route in order to get the loader function. If it was just _available_ that would be a massive simplification.

**Reply by akrynski-1** [reply]
This looks nice, having loader available pretty much solves the issue for me I think ❤

I had some use case where I had to distinguish between prefetch or regular load but now I don't remember what it was, I don't think this is very common use case so having it done with a context like this should probably be enough.

**Comment by maxprilutskiy** [comment]
This is a great addition, but I wonder if it could be complemented by the approach discussed in #13299?

The `onPrefetch` callback is flexible and explicit, but it requires **every link** to a route to implement the prefetch logic. This violates DRY and scatters data fetching concerns across multiple components instead of keeping them colocated with the route module.

For example:
```tsx
// With onPrefetch - repetitive across codebase
<Link to="/notes" prefetch="intent" onPrefetch={() => client.query(...)}>Notes</Link>
// Must be repeated in every component that links to /notes
```

vs

```tsx
// With clientLoader.prefetch - declared once in route
// routes/notes.tsx
export async function clientLoader() { ... }
clientLoader.prefetch = true;

// Anywhere in app - just works
<Link to="/notes" prefetch="intent">Notes</Link>
```

**Could both approaches coexist?** Route-level `clientLoader.prefetch` for the common case, and `onPrefetch` for edge cases requiring custom logic or dynamic parameters?

This would follow the existing pattern of `clientLoader.hydrate` and keep data concerns colocated with route modules per the React Router philosophy.

---

**Alternatively, we might not need this API at all** if React Router exposed prefetched data in a way that libraries like Apollo Client could access it.

The core issue is that [`useMatches()`](https://api.reactrouter.com/v7/functions/react_router.useMatches.html) only returns **matched routes**, not prefetched ones. When you hover a link with `prefetch="intent"`, the data is fetched and cached by the browser, but it never appears in `RouterState.loaderData` until you actually navigate.

This causes problems for cache-aware data libraries. For example, Apollo Client's [`ApolloHydrationHelper`](https://github.com/apollographql/apollo-client-integrations/blob/main/packages/react-router/src/ApolloHydrationHelper.tsx#L11) uses `useMatches()` to hydrate loader data into Apollo's cache:

```tsx
const matches = useMatches(); // ❌ Only contains matched routes, not prefetched
for (const match of matches) {
  // Hydrate match.data into Apollo cache
}
```

Since prefetched routes aren't in `matches`, their data never gets hydrated into Apollo's cache. When you navigate, the loader runs again (hitting the browser HTTP cache, but not Apollo's cache), causing components using `useReadQuery()` or `useSuspenseFragment()` to suspend and show loading states despite the data being prefetched.

**If React Router exposed prefetched loader data** (via `useMatches()`, a new `usePrefetchedRoutes()` hook, or adding it to `RouterState`), libraries could hydrate it proactively without requiring `onPrefetch` callbacks on every link.

Am I missing something here? Could we make this work, or is there an architectural reason why prefetched data can't be exposed in the router state?

**Reply by sergiodxa** [reply]
If it's the route the one who knows if it must prefetched, then this would required RR to download the JS of any linked route on the page to know if it should prefetch or not

This can cause more JS to be downloaded just for rendering a link

And because there's no other way to know if the linked route must be prefetched or not then there's no way to prevent this

The onPrefetch method while not DRY let the caller know if it should prefetch

This not only gives you more control, maybe you want to prefetch if coming from A but not if coming from B, it lets the router avoid having to download extra JS to know if it should prefetch or not

**Reply by maxprilutskiy** [reply]
@sergiodxa That's a great point about code splitting - I hadn't considered that the router would need to download route modules just to check `clientLoader.prefetch = true`. That would indeed defeat the purpose.

However, my second point still stands: **could React Router expose prefetched loader data in `RouterState`?** This would solve the cache hydration problem without requiring `onPrefetch` callbacks.

Currently, when `<Link prefetch="intent">` fetches data, it goes into the browser's HTTP cache but never appears in `RouterState.loaderData` until navigation completes. This means:

1. Data is prefetched ✅
2. Loader runs on navigation (but hits HTTP cache - fast) ✅
3. Loader returns data ✅
4. **Only NOW** does it appear in `RouterState.loaderData` and `useMatches()` ❌

By the time cache-aware libraries (Apollo, React Query, etc.) can see the data via `useMatches()`, the navigation is already complete. They've missed the opportunity to hydrate the prefetched data proactively.

**Could prefetched loader responses be added to `RouterState` immediately after prefetch completes?** Something like:

```tsx
interface RouterState {
  matches: AgnosticDataRouteMatch[];  // Current matched routes
  loaderData: RouteData;              // Data for matched routes
  prefetchedData: RouteData;          // NEW: Data for prefetched routes
  // ... other fields
}
```

Or expose a `usePrefetchedRoutes()` hook that returns prefetched data? (Though I'm honestly unsure if that's even possible.)

This would let libraries like Apollo hydrate prefetched data without needing `onPrefetch` on every link:

```tsx
// ApolloHydrationHelper.tsx
const matches = useMatches();
const prefetched = usePrefetchedRoutes(); // NEW

// Hydrate both matched AND prefetched data
[...matches, ...prefetched].forEach(route => {
  hydrateIntoCache(route.data);
});
```

This approach:
- ✅ No code splitting issues (data is already fetched)
- ✅ No `onPrefetch` needed on every link
- ✅ Libraries can proactively hydrate prefetched data
- ✅ Works with the existing `prefetch="intent"` behavior

**Regarding the code-splitting concern:** Could this be solved at build time? The framework could generate a manifest mapping routes to their prefetch needs, avoiding runtime JS downloads. But I understand if that's too complex - the `RouterState` exposure would still solve the hydration issue.

Thoughts?

**Reply by sergiodxa** [reply]
To have access to the data, RR would need to prefetch it itself, right now it relies on `<link rel="preload">` HTML tags to do the prefetch, that's why the browser does the fetch, but RR doesn't have access to it.

**Reply by brophdawg11** [reply]
>  Could this be solved at build time?

One of the big things this proposal is after is bringing a Framework Mode-only feature down into Data Mode where we don't own the build step.  That's why, at first glance, this might seem a bit more lower level.  We want it to be flexible so it can be used in _any_ way in Data Mode.  Framework mode can always build upon that in the future but I don't think that's the primary goal of this RFC.

I think the ability to [call the loader](https://github.com/remix-run/react-router/discussions/12375#discussioncomment-14616474) and also [leverage `route.handle`](https://github.com/remix-run/react-router/discussions/12375#discussioncomment-14180792) will let you build something close to what you're after.

**Reply by maxprilutskiy** [reply]
@brophdawg11 @sergiodxa this makes a lot of sense to me now. Thank you guys!

**Reply by ariesclark** [reply]
> If it's the route the one who knows if it must prefetched, then this would required RR to download the JS of any linked route on the page to know if it should prefetch or not

Isn't this solved by     ``unstable_splitRouteModules: true``?

**Reply by sergiodxa** [reply]
That splits the route modules in different files, but the router will still need to download the file that knows if it should prefetch or not, just to be able to prefetch

**Comment by buzinas** [comment]
Just to make sure I don't spend a lot of time for nothing: is this `accepting-prs` still valid? @brophdawg11 

I'd like to work on this feature, as it would be very aligned to the work I'm currently doing on our app.

**Reply by brophdawg11** [reply]
yep!  Feel free to take a stab and open a PR.  Feel free to go the simple argument-less route in the original description, or if you want to look into allowing calling the route loader like we discussed in https://github.com/remix-run/react-router/discussions/12375#discussioncomment-14616474 that might be neat to see how complex that would be.

Thanks!

**Comment by jasperfurniss** [comment]
Hi @buzinas, are you still working on this? If not, I'd like to pick it up.

I was noodling on the scope a little....based on the original proposal:

- Add `unstable_onPrefetch` to `Link` (and so `NavLink`) with no arguments
- Call it when the existing `prefetch` behavior triggers: `intent` (hover, focus, or touch after the 100ms delay), `render`, and `viewport`
- Make it work in Data Mode too, where `prefetch` currently does nothing. Without `unstable_onPrefetch`, Data Mode behavior stays the same. In Framework Mode it runs alongside the built-in `<link rel="prefetch">` tags
- Leave calling the route loader for a follow-up

One question: should it fire every time intent re-triggers (hovering away and back, or scrolling out of view ..and back into view) and leave deduping to the app's cache, or only once per link?

cc: @brophdawg11
